// @langchain/langgraph/dist/pregel/call.js
import { RunnableSequence, } from "@langchain/core/runnables";
import { AsyncLocalStorageProviderSingleton } from "@langchain/core/singletons";
import { CONFIG_KEY_CALL, END, PREVIOUS, RETURN, TAG_HIDDEN, } from "../constants.js";
import { ChannelWrite, PASSTHROUGH } from "./write.js";
import { RunnableCallable } from "../utils.js";
import { isEntrypointFinal, } from "../func/types.js";
/**
 * Wraps a user function in a Runnable that writes the returned value to the RETURN channel.
 */
export function getRunnableForFunc(name, func) {
    const run = new RunnableCallable({
        func: (input) => func(...input),
        name,
        trace: false,
        recurse: false,
    });
    return new RunnableSequence({
        name,
        first: run,
        last: new ChannelWrite([{ channel: RETURN, value: PASSTHROUGH }], [TAG_HIDDEN]),
    });
}
export function getRunnableForEntrypoint(name, func) {
    const run = new RunnableCallable({
        func: (input, config) => {
            return func(input, config);
        },
        name,
        trace: false,
        recurse: false,
    });
    return new RunnableSequence({
        name,
        first: run,
        middle: [
            new ChannelWrite([
                {
                    channel: END,
                    value: PASSTHROUGH,
                    mapper: new RunnableCallable({
                        func: (value) => (isEntrypointFinal(value) ? value.value : value),
                    }),
                },
            ], [TAG_HIDDEN]),
            new ChannelWrite([
                {
                    channel: PREVIOUS,
                    value: PASSTHROUGH,
                    mapper: new RunnableCallable({
                        func: (value) => {
                            return isEntrypointFinal(value) ? value.save : value;
                        },
                    }),
                },
            ]),
        ],
        last: new RunnableCallable({
            func: (final) => isEntrypointFinal(final) ? final.value : final,
        }),
    });
}
export function call({ func, name, retry }, ...args) {
    const config = AsyncLocalStorageProviderSingleton.getRunnableConfig();
    if (typeof config.configurable?.[CONFIG_KEY_CALL] === "function") {
        return config.configurable[CONFIG_KEY_CALL](func, name, args, {
            retry,
            callbacks: config.callbacks,
        });
    }
    throw new Error("Async local storage not initialized. Please call initializeAsyncLocalStorageSingleton() before using this function.");
}
