// @langchain/langgraph-sdk/dist/react-ui/server/server.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.typedUi = void 0;
const uuid_1 = require("uuid");
const typedUi = (config) => {
    let items = [];
    const runId = config.metadata?.run_id ?? config.runId;
    if (!runId)
        throw new Error("run_id is required");
    const metadata = {
        ...config.metadata,
        tags: config.tags,
        name: config.runName,
        run_id: runId,
    };
    const handlePush = (message, options) => {
        const evt = {
            type: "ui",
            id: message?.id ?? (0, uuid_1.v4)(),
            name: message?.name,
            props: message?.props,
            metadata: {
                ...metadata,
                ...message?.metadata,
                ...(options?.message ? { message_id: options.message.id } : null),
            },
        };
        items.push(evt);
        config.writer?.(evt);
        return evt;
    };
    const handleDelete = (id) => {
        const evt = { type: "remove-ui", id };
        items.push(evt);
        config.writer?.(evt);
        return evt;
    };
    return { push: handlePush, delete: handleDelete, items };
};
exports.typedUi = typedUi;
